### PR TITLE
Fix nasty bug in DFT_deconv

### PR DESCRIPTION
**Pull request by BjoernLudwigPTB**
We expect, that there is an error in `DFT_deconv` as one of the formulas seems to differ from what is expressed in the paper. An industry partner hinted us towards an error in

```python
    RH = np.r_[
        (-rY * rH ** 2 + rY * iH ** 2 - 2 * iY * iH * rH) / norm ** 2,
        (iY * rH ** 2 - iH * iH ** 2 - 2 * rY * rH * iH) / norm ** 2,
    ]
```
where it should be 
```python
    RH = np.r_[
        (-rY * rH ** 2 + rY * iH ** 2 - 2 * iY * iH * rH) / norm ** 2,
        (iY * rH ** 2 - iY * iH ** 2 - 2 * rY * rH * iH) / norm ** 2,
    ]
```
Spot the error?

Finally we succeeded in showing the error here in this PR and fixing it in #221. We already merged the slow-tag to execute the fully randomized test only on nightly runs. After finishing the review we can then finally merge #221 into this one here and then release it by merging into master.

### CODE REVIEW

File: test/test_DFT.py [hunk 1]
```diff
@@ -159,25 +116,90 @@ def x_Ux_and_window(draw: Callable, ux_type: Optional[Union[np.ndarray, float]]
...
+    """Test GUM_DFT and GUM_iDFT with full covariance matrix"""
+    x, ux = multisine_testsignal
+    ux = np.ones_like(x) * 0.01 ** 2
+    cx = corrmatrix(0.95, len(x))
+    Ux = np.diag(ux)
+    Ux = Ux.dot(cx.dot(Ux))
+    X, UX = GUM_DFT(x, Ux)
+    xh, Uxh = GUM_iDFT(X, UX)
+    assert_almost_equal(np.max(np.abs(x - xh)), 0)
+    assert_almost_equal(np.max(Ux - Uxh), 0)
```
Comment (mgrub) [inline test/test_DFT.py:200]: What was the reasoning to not put `test_compose_DFT_and_iDFT_with_full_covariance` inside `class TestDFT`?

File: test/test_DFT_deconv.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    h_complex = hypothesis.draw(nonzero_complex_vector(length=n))
+    h = np.r_[h_complex.real, h_complex.imag]
+    uh = covariance_scale_minimizer * hypothesis.draw(
+        hypothesis_covariance_matrix_for_complex_vectors(n)
+    )
+    x_deconv, u_deconv = DFT_deconv(H=h, Y=y, UH=uh, UY=uy)
+    n_monte_carlo_runs = 40000
+    y_mc = stats.multivariate_normal.rvs(mean=y, cov=uy, size=n_monte_carlo_runs)
+    h_mc = stats.multivariate_normal.rvs(mean=h, cov=uh, size=n_monte_carlo_runs)
+    real_complex_divider_index = 2 * n // 2
```
Comment (mgrub) [inline test/test_DFT_deconv.py:38]: I am curious: In which corner cases `n != real_complex_divider_index` ?

File: test/test_DFT_deconv.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    y_mcs_divided_by_h_mcs = np.concatenate(
+        (
+            np.real(y_mcs_divided_by_h_mcs_complex),
+            np.imag(y_mcs_divided_by_h_mcs_complex),
+        ),
+        axis=1,
+    )
+    y_divided_by_h_mc_mean = np.mean(y_mcs_divided_by_h_mcs, axis=0)
+    y_divided_by_h_mc_cov = np.cov(y_mcs_divided_by_h_mcs, rowvar=False)
+    assert_allclose(x_deconv + 1, y_divided_by_h_mc_mean + 1, rtol=3e-4, atol=4e-7)
```
Comment (mgrub) [inline test/test_DFT_deconv.py:57]: Is the `+1` necessary to avoid issues with relative tolerances?

**Review by mgrub — COMMENTED** [review]
That is quite a lot changes caused by a single character-typo ;-)

Very nice and clean work - I am looking forward to our in-depth review.

File: test/test_DFT.py (hunk 1 above)
Comment (BjoernLudwigPTB) [reply test/test_DFT.py:200]: Originally it was part of the class under the name of `test_DFT_iDFT_fullcov`. I touched this method, because in an intermediate step I made use of the `corrmatrix` function. Then Codacy complained about it being a method unnecessarily since there is no use of `self`. That's why I refactored it to being a function outside of the class. I think, that I already did refactor the whole of the class in the `prepare_major_release_2.0.0` . Here I just wanted to satisfy Codacy with minimal effort. The renaming still seemed reasonable to me.

File: test/test_DFT_deconv.py (hunk 2 above)
Comment (BjoernLudwigPTB) [reply test/test_DFT_deconv.py:38]: It will be different for odd n. I took this snippet equivalently from the original functions indexing: https://github.com/PTB-M4D/PyDynamic/blob/befb588c45c29a6aa746379eddc5b9653acdd653/PyDynamic/uncertainty/propagate_DFT.py#L971.

File: test/test_DFT_deconv.py (hunk 3 above)
Comment (BjoernLudwigPTB) [reply test/test_DFT_deconv.py:57]: Correct, we shift away from zero in this comparison to avoid issues with the formula that `assert_allclose` uses: `atol + rtol * abs(desired)` which in my view renders less informative for `desired` being the zero vector.

File: test/conftest.py [hunk 4]
```diff
@@ -1,13 +1,250 @@
...
+    def _create_corrmatrix(
+        rho: float, Nx: int, nu: Optional[float] = 0.5, phi: Optional[float] = 0.3
+    ) -> np.ndarray:
+        """Additional helper function to create a correlation matrix"""
+        corrmat = np.zeros((Nx, Nx))
+        if rho > 1:
+            raise ValueError("Correlation scalar should be less than one.")
+
+        for k in range(1, Nx):
+            corrmat += np.diag(np.ones(Nx - k) * rho ** (phi * k ** nu), k)
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply test/conftest.py:245]: Oh yes indeed. I only replace refactored this function to make use of it in more than only the original use case (which was the `DFT_iDFT` thing). I would suggest to open a new issue for this suggestion though to save me from the hour of implementing it properly. In fact at the moment we create valid covariance matrices in several ways now for different purposes where I am not sure, if all of these procedures are equally well thought through. The one making use of this function for instance seems pretty blown up for simply constructing a reasonable (and still fixed) covariance matrix.

File: test/test_DFT_deconv.py (hunk 2 above)
Comment (BjoernLudwigPTB) [reply test/test_DFT_deconv.py:40]: I love your humble question: "In which 'corner cases' do we have 2/2 != 1?" Made my day!

Comment (BjoernLudwigPTB) [reply test/test_DFT_deconv.py]: Resolved with a12d42d.

File: test/conftest.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply test/conftest.py:245]: Resolved via opening issue #225 .

File: test/test_DFT.py (hunk 1 above)
Comment (BjoernLudwigPTB) [reply test/test_DFT.py:200]: Resolved and approved via telephone just now.

File: test/test_DFT_deconv.py (hunk 3 above)
Comment (BjoernLudwigPTB) [reply test/test_DFT_deconv.py:58]: Resolved and approved via telephone just now.
